starky/src/recursive_verifier.rs

```diff
@@ -0,0 +1,261 @@
+use itertools::Itertools;
+use plonky2::field::extension_field::Extendable;
+use plonky2::field::field_types::Field;
+use plonky2::fri::witness_util::set_fri_proof_target;
+use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::iop::witness::Witness;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::config::{AlgebraicHasher, GenericConfig};
+use plonky2::util::reducing::ReducingFactorTarget;
+
+use crate::config::StarkConfig;
+use crate::constraint_consumer::RecursiveConstraintConsumer;
+use crate::proof::{
+    StarkOpeningSetTarget, StarkProof, StarkProofChallengesTarget, StarkProofTarget,
+    StarkProofWithPublicInputs, StarkProofWithPublicInputsTarget,
+};
+use crate::stark::Stark;
+use crate::vars::StarkEvaluationTargets;
+
+pub fn recursively_verify_stark_proof<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    stark: S,
+    proof_with_pis: StarkProofWithPublicInputsTarget<D>,
+    inner_config: &StarkConfig,
+) where
+    C::Hasher: AlgebraicHasher<F>,
+    [(); S::COLUMNS]:,
+    [(); S::PUBLIC_INPUTS]:,
+{
+    assert_eq!(proof_with_pis.public_inputs.len(), S::PUBLIC_INPUTS);
+    let degree_bits = proof_with_pis.proof.recover_degree_bits(inner_config);
+    let challenges = proof_with_pis.get_challenges::<F, C>(builder, inner_config, degree_bits);
+
+    recursively_verify_stark_proof_with_challenges::<F, C, S, D>(
+        builder,
+        stark,
+        proof_with_pis,
+        challenges,
+        inner_config,
+        degree_bits,
+    );
+}
+
+/// Recursively verifies an inner proof.
+fn recursively_verify_stark_proof_with_challenges<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    S: Stark<F, D>,
+    const D: usize,
+>(
+    builder: &mut CircuitBuilder<F, D>,
+    stark: S,
+    proof_with_pis: StarkProofWithPublicInputsTarget<D>,
+    challenges: StarkProofChallengesTarget<D>,
+    inner_config: &StarkConfig,
+    degree_bits: usize,
+) where
+    C::Hasher: AlgebraicHasher<F>,
+    [(); S::COLUMNS]:,
+    [(); S::PUBLIC_INPUTS]:,
+{
+    let one = builder.one_extension();
+
+    let StarkProofWithPublicInputsTarget {
+        proof,
+        public_inputs,
+    } = proof_with_pis;
+    let local_values = &proof.openings.local_values;
+    let next_values = &proof.openings.local_values;
```

Comment:
`.local_values` -> `.next_values`

Reply:
Oh seems these are overwritten by the line below?

Reply:
Oops yeah don't know what these 2 lines were for. They were also in the native verifier. I've removed them.